src/Components/WebView/WebView/src/IpcReceiver.cs

```diff
@@ -0,0 +1,119 @@
+using System;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components.Web;
+using Microsoft.JSInterop.Infrastructure;
+
+// Sync vs Async APIs for this.
+// This mainly depends on the underlying support for the browser. Assuming that there is no synchronous API
+// communication is safer, since it's not guaranteed.
+// In that scenario, some APIs need to expose the async nature of the communication. That happens when some
+// component like the renderer needs to know the results of the operation. For example when updating the UI
+// since more code needs to execute afterwards.
+// In other cases like when we try to attach a component to the document, we don't necessarily need to do that
+// since we only care about errors that might happen while attaching the component and the renderer doesn't
+// necessarily need to know about those if we are terminating the component/host as a result.
+// If we decide we need to expose the async nature of the communication channel, then we will need to keep track
+// of all the message pairs/completions across the IPC channel.
+namespace Microsoft.AspNetCore.Components.WebView
+{
+    // These are all the messages .NET Host needs to know how to receive from JS
+
+    // This class is a "Proxy" or "front-controller" for the incoming messages from the Browser via the transport channel.
+    // It receives messages on OnMessageReceived, interprets the payload and dispatches them to the appropriate method
+    internal class IpcReceiver
+    {
+        private readonly WebViewManager _manager;
+
+        public IpcReceiver(WebViewManager manager)
+        {
+            _manager = manager;
+        }
```

Comment:
I think we should reconsider the direction of this dependency.

The WebViewManager creates and manages the IPCReceiver, but then the receiver calls back to the manager and both reference each other.

I think we might be better of if we follow an alternative approach:
1) Handle initialization outside the IpcReceiver.
2) Instead of dispatching the "AttachToPage" we expose an "Initialize" event that the manager subscribes to from outside.

That way all interactions always go Manager -> (IPCXXX) and there is no coupling both ways.

Reply:
Absolutely, sounds good to me! I agree it's a bit unexpected that it relies on an `internal` method at the moment. The only drawback that concerns me about using an event is that it's hard to deal with asynchrony and exceptions, but we can figure something out.